pkg/webhook/mutation.go

```diff
@@ -162,9 +170,56 @@ func (h *mutationHandler) Handle(ctx context.Context, req admission.Request) adm
 }
 
 func (h *mutationHandler) mutateRequest(ctx context.Context, req admission.Request) (admission.Response, error) {
-	// TODO: place mutation logic here
-	mutatedObject := req.Object
-	resp := admission.PatchResponseFromRaw(req.Object.Raw, mutatedObject.Raw)
+
+	okResp := admission.Allowed("")
+	ns := &corev1.Namespace{}
+
+	// if the object being mutated is a namespace itself, we use it as namespace
+	if req.Kind.Kind == namespaceKind && req.Kind.Group == "" {
+		req.Namespace = ""
+		obj, _, err := deserializer.Decode(req.Object.Raw, nil, &corev1.Namespace{})
+		if err != nil {
+			return admission.Errored(int32(http.StatusInternalServerError), err), nil
+		}
+		ok := false
+		ns, ok = obj.(*corev1.Namespace)
+		if !ok {
+			return admission.Errored(int32(http.StatusInternalServerError), errors.New("failed to cast namespace object")), nil
+		}
+	} else {
+		if err := h.client.Get(ctx, types.NamespacedName{Name: req.AdmissionRequest.Namespace}, ns); err != nil {
+			if !k8serrors.IsNotFound(err) {
+				log.Info("Namespace not found", "name", req.AdmissionRequest.Namespace)
+				return okResp, nil
+			}
+			// bypass cached client and ask api-server directly
+			err = h.reader.Get(ctx, types.NamespacedName{Name: req.AdmissionRequest.Namespace}, ns)
+			if err != nil {
+				log.Info("Namespace not found", "name", req.AdmissionRequest.Namespace)
+				return okResp, nil
```

Comment:
I am less certain we want to return an "ok" response here. The most likely reason for a missing namespace is that there is some lag in eventual consistency. IMO the correct response to that is to fail and let the client retry.

Reply:
Should we return both of the above as 500 errors?